File: 프로그래머스/2/87946. 피로도/피로도.java
class Solution {
    int answer = 0;
    boolean[] visited;
    public int solution(int k, int[][] dungeons) {
       visited = new boolean[dungeons.length];
        
        dfs(k,0,dungeons);
        return answer;
        
        
    }
    void dfs(int k, int depth, int[][] dungeons) {
        for(int i = 0; i < dungeons.length;i++) {
            if(!visited[i] && k >= dungeons[i][0]) {
                visited[i] = true;
                dfs(k - dungeons[i][1], depth+1, dungeons);
                visited[i] = false;
            }
        }
        answer = Math.max(answer,depth);
    }
    
    
}

File: 프로그래머스/2/87946. 피로도/피로도.js
function dfs(curK, cnt, dungeons, visited) {
    let max = cnt;
    for (let i = 0; i < dungeons.length; i++) {
        if (curK >= dungeons[i][0] && !visited.has(i)) {
            visited.add(i);
            max = Math.max(max, dfs(curK - dungeons[i][1], cnt + 1 ,dungeons, visited))
            visited.delete(i);
        }
    }
    return max;
}

function solution(k, dungeons) {
    var answer = 0;
    const visited = new Set();
    answer = dfs(k,0,dungeons,visited);
    return answer;
}